File: android/core/src/main/kotlin/io/mynote/core/FractionalIndex.kt
package io.mynote.core

/**
 * Ordering keys for blocks and notes — the Kotlin twin of `FractionalIndex.swift`.
 *
 * Storing an integer position would mean rewriting every row below an insert, and
 * two offline devices inserting at the same spot would collide. A fractional index
 * generates a *string* strictly between its neighbours, so an insert touches one
 * row and concurrent inserts merely interleave.
 *
 * Invariant: **a generated key never ends in the lowest digit ('0')**. Without it
 * there is no key below `"0"` (`"00"` sorts *after* `"0"`), so inserting at the top
 * repeatedly would eventually have no answer.
 */
object FractionalIndex {
    private const val DIGITS = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
    private val BASE = DIGITS.length

    /** Depth guard: a bad call must not hang the editor. */
    private const val MAX_DEPTH = 512

    /**
     * A key strictly between [a] and [b]. `null` means "no neighbour on that side".
     *
     * Requires `a < b` when both are given.
     */
    fun between(a: String?, b: String?): String {
        // Invalid input (`a >= b`, or a `b` violating the trailing-digit
        // invariant) yields a best-effort key rather than an exception. The
        // server rejects malformed order keys at the boundary, so reaching this
        // is already a bug elsewhere — and mis-ordering one block beats crashing
        // the editor while someone is writing. Swift behaves identically.
        val lower = a ?: ""
        var upper: String? = b
        val result = StringBuilder()
        var i = 0

        while (i < MAX_DEPTH) {
            val ca: Int? = if (i < lower.length) DIGITS.indexOf(lower[i]).takeIf { it >= 0 } else null

            // BASE means "unconstrained above"; 0 means b ran out or holds the
            // lowest digit, and either way there is no room at this position.
            val cb: Int = upper?.let { up ->
                if (i < up.length) DIGITS.indexOf(up[i]).coerceAtLeast(0) else 0
            } ?: BASE

            if (ca == null) {
                // `a` ran out: anything appended already sorts after it, so we
                // only have to stay below `b`.
                if (cb > 1) {
                    result.append(DIGITS[cb / 2])   // >= 1, never a trailing '0'
                    return result.toString()
                }
                result.append(DIGITS[0])
                // Placing '0' under b's '1' puts us strictly below b already.
                if (cb == 1) upper = null
                i++
                continue
            }

            if (cb - ca > 1) {
                result.append(DIGITS[(ca + cb) / 2])   // > ca, never '0'
                return result.toString()
            }

            result.append(DIGITS[ca])
            if (ca < cb) upper = null   // diverged below b; suffix is now free
            i++
        }

        result.append(DIGITS[BASE / 2])
        return result.toString()
    }

    fun sequence(count: Int): List<String> {
        val keys = mutableListOf<String>()
        var previous: String? = null
        repeat(count) {
            val key = between(previous, null)
            keys += key
            previous = key
        }
        return keys
    }
}

File: ios/MyNoteCore/Sources/MyNoteCore/FractionalIndex.swift
import Foundation

/// Ordering keys for blocks and notes.
///
/// Storing an integer `position` would mean rewriting every row below an insert,
/// and two offline devices inserting at the same spot would collide. A fractional
/// index instead generates a *string* strictly between its neighbours, so an
/// insert touches exactly one row and concurrent inserts merely interleave.
///
/// Invariant: **a generated key never ends in the lowest digit ('0')**.
/// Without it there is no key below `"0"` — `"00"` sorts *after* `"0"` — and
/// `between(nil, "0")` has no answer, so repeatedly inserting at the top would
/// eventually wedge. Keeping the last digit non-zero guarantees there is always
/// room below any existing key.
public enum FractionalIndex {
    static let digits = Array("0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz")
    static let base = digits.count

    /// Depth guard. Keys grow by one character per exhausted gap, so real use
    /// stays far below this; hitting it means a caller passed `a >= b`, and
    /// returning a usable key beats hanging the editor.
    static let maxDepth = 512

    private static func index(of ch: Character) -> Int? {
        digits.firstIndex(of: ch)
    }

    /// A key strictly between `a` and `b`. `nil` means "no neighbour on that side".
    ///
    /// Precondition: if both are given, `a < b`.
    /// Invalid input (`a >= b`, or a `b` that violates the trailing-digit
    /// invariant) yields a best-effort key rather than a crash or a hang. The
    /// server rejects malformed order keys at the boundary, so reaching this is
    /// already a bug elsewhere — and mis-ordering one block beats killing the
    /// editor while someone is writing. Kotlin behaves identically.
    public static func between(_ a: String?, _ b: String?) -> String {
        let lower = Array(a ?? "")
        var upper: [Character]? = b.map(Array.init)
        var result = ""
        var i = 0

        while i < maxDepth {
            let ca: Int? = i < lower.count ? index(of: lower[i]) : nil

            // `base` means "unconstrained above"; 0 means b has run out or holds
            // the lowest digit, and either way there is no room at this position.
            let cb: Int
            if let up = upper {
                cb = i < up.count ? (index(of: up[i]) ?? 0) : 0
            } else {
                cb = base
            }

            guard let ca else {
                // `a` has run out: anything we append already sorts after it, so
                // we only have to stay below `b`.
                if cb > 1 {
                    result.append(digits[cb / 2])   // >= 1, so never a trailing '0'
                    return result
                }
                // cb is 0 or 1: no room here, so step down a level.
                result.append(digits[0])
                // Placing '0' under b's '1' puts us strictly below b already,
                // which frees the rest of the suffix.
                if cb == 1 { upper = nil }
                i += 1
                continue
            }

            if cb - ca > 1 {
                result.append(digits[(ca + cb) / 2])   // > ca, so never '0'
                return result
            }

            result.append(digits[ca])
            if ca < cb { upper = nil }   // diverged below `b`; suffix is now free
            i += 1
        }

        // Unreachable for valid input; keeps a bad call from hanging.
        result.append(digits[base / 2])
        return result
    }

    /// Evenly spaced keys for an initial batch.
    public static func sequence(count: Int) -> [String] {
        guard count > 0 else { return [] }
        var keys: [String] = []
        var previous: String? = nil
        for _ in 0..<count {
            let key = between(previous, nil)
            keys.append(key)
            previous = key
        }
        return keys
    }
}
